File: problem120.cpp
#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        // DP problem 
        // Base case 
        if(triangle.size() == 1){
            return triangle[0][0];
        }

        // Goes from bottom to top
        // Update number from bottom row, and choose the min number from the option

        for(int row = triangle.size()-1; row > 0; row--){
            // Goes through all number in current row
            for(int index = 0; index < triangle[row].size()-1; index++){
                // Update the next row
                triangle[row-1][index] += min(triangle[row][index], triangle[row][index+1]);
            }
        }
        return triangle[0][0];
    }
};

File: problem120.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        # Go through all the rows 
        # Choose smaller number from index i or i + 1

        # Base case
        if len(triangle) == 1:
            return triangle[0][0]

        # Dynamic programming 
        # Changable variables for sum of numbers and index 

        for row in range(len(triangle)-1, 0, -1):
            # Start from bottom to top
            # Update second to last row 
            # For each node, add min number of two numbers
            for col in range(len(triangle[row])-1):
                # Update row above
                triangle[row-1][col] += min(triangle[row][col], triangle[row][col+1])
                
        return triangle[0][0]